File: app/src/main/java/com/example/pam_project/repositories/lists/RoomListsRepository.java
package com.example.pam_project.repositories.lists;

import com.example.pam_project.database.lists.ListDao;
import com.example.pam_project.database.lists.ListEntity;
import com.example.pam_project.database.lists.ListMapper;
import com.example.pam_project.features.lists.list.ListInformation;

import io.reactivex.Completable;
import io.reactivex.Flowable;

public class RoomListsRepository implements ListsRepository {
    private final ListDao listDao;
    private final ListMapper mapper;

    public RoomListsRepository(final ListDao listDao, final ListMapper mapper) {
        this.listDao = listDao;
        this.mapper = mapper;
    }

    @Override
    public Flowable<ListInformation> getList(final long listId) {
        return listDao.getListById(listId).map(mapper::toModel);
    }

    @Override
    public Flowable<ListInformation> getListWithTasks(final long listId) {
        return listDao.getListsWithTasks(listId).map(mapper::toListWithTasksModel);
    }

    @Override
    public Completable insertList(final String name, final long categoryId) {
        ListEntity listEntity = new ListEntity(name, categoryId);
        return Completable.fromAction(() -> listDao.insertList(listEntity));
    }

    @Override
    public Completable updateList(final long id, final String name, final long categoryId) {
        ListEntity listEntity = new ListEntity(id, name, categoryId);
        return Completable.fromAction(() -> listDao.updateList(listEntity));
    }

    @Override
    public Completable deleteList(long id) {
        ListEntity listEntity = listDao.getListById(id).blockingFirst();
        return Completable.fromAction(() -> listDao.deleteList(listEntity));
    }
}

File: app/src/main/java/com/example/pam_project/repositories/lists/RoomListsRepository.kt
package com.example.pam_project.repositories.lists

import com.example.pam_project.database.lists.ListDao
import com.example.pam_project.database.lists.ListEntity
import com.example.pam_project.database.lists.ListMapper
import com.example.pam_project.database.relationships.ListsWithTasks
import com.example.pam_project.features.lists.list.ListInformation
import io.reactivex.Completable
import io.reactivex.Flowable

class RoomListsRepository(private val listDao: ListDao?, private val mapper: ListMapper?) : ListsRepository {
    override fun getList(listId: Long): Flowable<ListInformation?> {
        return listDao!!.getListById(listId).map { entity: ListEntity? -> mapper!!.toModel(entity) }
    }

    override fun getListWithTasks(listId: Long): Flowable<ListInformation?> {
        return listDao!!.getListsWithTasks(listId).map { entity: ListsWithTasks? -> mapper!!.toListWithTasksModel(entity) }
    }

    override fun insertList(name: String, categoryId: Long): Completable {
        val listEntity = ListEntity(name, categoryId)
        return Completable.fromAction { listDao!!.insertList(listEntity) }
    }

    override fun updateList(id: Long, name: String, categoryId: Long): Completable {
        val listEntity = ListEntity(id, name, categoryId)
        return Completable.fromAction { listDao!!.updateList(listEntity) }
    }

    override fun deleteList(id: Long): Completable {
        val listEntity = listDao!!.getListById(id).blockingFirst()
        return Completable.fromAction { listDao.deleteList(listEntity) }
    }
}